resample/bootstrap.py

```diff
@@ -0,0 +1,285 @@
+
+    alpha = 1 - cl
+    thetas = bootstrap(fn, sample, **kwds)
+
+    if ci_method == "percentile":
+        return _confidence_interval_percentile(thetas, alpha / 2)
+
+    theta = fn(sample)
+
+    if ci_method == "student":
+        return _confidence_interval_studentized(theta, thetas, alpha / 2)
+
+    if ci_method == "bca":
+        j_thetas = _jackknife(fn, sample)
+        return _confidence_interval_bca(theta, thetas, j_thetas, alpha / 2)
+
+    raise ValueError(
+        "ci_method must be 'percentile', 'student', or 'bca', but "
+        f"'{ci_method}' was supplied"
+    )
+
+
+def _resample_stratified(sample, size, method, strata, rng):
+    # call resample on sub-samples and merge the replicas
+    sub_samples = [sample[strata == x] for x in np.unique(strata)]
+    for sub_replicas in zip(
+        *[resample(s, size, method=method, random_state=rng) for s in sub_samples]
+    ):
+        yield np.concatenate(sub_replicas, axis=0)
+
+
+def _resample_ordinary(
+    sample: np.ndarray, size: int, rng: np.random.Generator,
+) -> np.ndarray:
+    # i.i.d. sampling from empirical cumulative distribution of sample
+    n = len(sample)
+    for _ in range(size):
+        yield rng.choice(sample, size=n, replace=True)
+
+
+def _resample_balanced(
+    sample: np.ndarray, size: int, rng: np.random.Generator,
+) -> np.ndarray:
+    # effectively computes a random permutation of `size` concatenated
+    # copies of `sample` and returns `size` equal chunks of that
+    n = len(sample)
+    indices = rng.permutation(n * size)
+    for i in range(size):
+        sel = indices[i * n : (i + 1) * n] % n
+        yield sample[sel]
+
+
+def _fit_parametric_family(dist: stats.rv_continuous, sample: np.ndarray) -> Tuple:
+    if dist == stats.multivariate_normal:
+        # has no fit method...
+        return (np.mean(sample, axis=0), np.cov(sample.T, ddof=1))
+
+    if dist == stats.t:
+        fit_kwd = {"fscale": 1}  # HD: I think this should not be fixed to 1
+    elif dist in {stats.f, stats.beta}:
+        fit_kwd = {"floc": 0, "fscale": 1}
+    elif dist in (stats.gamma, stats.lognorm, stats.invgauss, stats.pareto):
+        fit_kwd = {"floc": 0}
+    else:
+        fit_kwd = {}
+
+    return dist.fit(sample, **fit_kwd)
+
+
+def _resample_parametric(
+    sample: np.ndarray, size: int, dist, rng: np.random.Generator,
+) -> np.ndarray:
+    n = len(sample)
+
+    # fit parameters by maximum likelihood and sample from that
+    if dist == stats.poisson:
+        # - poisson has no fit method and there is no scale parameter
+        # - random number generation for poisson distribution in scipy seems to be buggy
+        mu = np.mean(sample)
+        for _ in range(size):
+            yield rng.poisson(mu, size=n)
+    else:
+        args = _fit_parametric_family(dist, sample)
+        dist = dist(*args)
+        for _ in range(size):
+            yield dist.rvs(size=n, random_state=rng)
+
+
+def _confidence_interval_percentile(
```

Comment:
Thanks for breaking this into helper functions, I think it improves readability / maintainability a lot

Reply:
Thanks!

I sometimes also write long functions, instead of factoring things out, like one should. You asked before what pylint does on top of flake8. pylint, for example, complained here about the original function being too long and having too many 'if's, and thus encouraged me to factor things out into smaller pieces.